File: test_buildings.py
import urllib.request
import urllib.parse
import json
import math

def haversine(lat1, lon1, lat2, lon2):
    R = 6371000
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlam = math.radians(lon2 - lon1)
    a = math.sin(dphi/2)**2 + math.cos(phi1)*math.cos(phi2)*math.sin(dlam/2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return R * c

def bearing(lat1, lon1, lat2, lon2):
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    dlam = math.radians(lon2 - lon1)
    y = math.sin(dlam) * math.cos(phi2)
    x = math.cos(phi1)*math.sin(phi2) - math.sin(phi1)*math.cos(phi2)*math.cos(dlam)
    theta = math.atan2(y, x)
    return (math.degrees(theta) + 360) % 360

lat = 40.7127
lon = -74.0134
radius = 250

query = f"""
[out:json];
(
  way["building"](around:{radius},{lat},{lon});
  relation["building"](around:{radius},{lat},{lon});
);
out center tags;
"""

url = "http://overpass-api.de/api/interpreter?data=" + urllib.parse.quote(query)
req = urllib.request.Request(url, headers={'User-Agent': 'FelloSiteChecker/1.0'})

angles = [{"dir": d, "angle": 0} for d in ['N','NNE','NE','ENE','E','ESE','SE','SSE','S','SSW','SW','WSW','W','WNW','NW','NNW']]

try:
    with urllib.request.urlopen(req) as response:
        data = json.loads(response.read().decode())
        applied = 0
        for el in data.get('elements', []):
            center = el.get('center')
            if not center: continue
            
            blat, blon = center['lat'], center['lon']
            dist = haversine(lat, lon, blat, blon)
            if dist < 5: continue
            
            b = bearing(lat, lon, blat, blon)
            
            h = 12
            tags = el.get('tags', {})
            if 'height' in tags:
                try: h = float(tags['height'])
                except: pass
            elif 'building:levels' in tags:
                try: h = float(tags['building:levels']) * 3.5
                except: pass
            
            elev_angle = math.degrees(math.atan2(h, dist))
            bin_index = round(b / 22.5) % 16
            
            if elev_angle > angles[bin_index]['angle']:
                angles[bin_index]['angle'] = round(elev_angle, 1)
                applied += 1
                
        print(f"Applied {applied} buildings")
        for a in angles:
            print(f"{a['dir']}: {a['angle']} degrees")
except Exception as e:
    print(e)

File: test_buildings.js
const fetch = require('node-fetch');

async function getBuildingObstructions(lat, lon, horizonAngles) {
  const radius = 250; // meters
  const query = `[out:json];(way["building"](around:${radius},${lat},${lon});relation["building"](around:${radius},${lat},${lon}););out center tags;`;
  const url = `https://overpass-api.de/api/interpreter?data=${encodeURIComponent(query)}`;
  
  try {
    const res = await fetch(url);
    const data = await res.json();
    if (!data.elements) return horizonAngles;

    const mergedAngles = JSON.parse(JSON.stringify(horizonAngles));
    let applied = 0;
    
    data.elements.forEach(el => {
      if (!el.center) return;
      const blat = el.center.lat, blon = el.center.lon;
      
      const R = 6371e3;
      const φ1 = lat * Math.PI/180, φ2 = blat * Math.PI/180;
      const Δφ = (blat-lat) * Math.PI/180, Δλ = (blon-lon) * Math.PI/180;
      const a = Math.sin(Δφ/2) * Math.sin(Δφ/2) + Math.cos(φ1) * Math.cos(φ2) * Math.sin(Δλ/2) * Math.sin(Δλ/2);
      const dist = R * (2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a)));

      if (dist < 5) return;

      const y = Math.sin(Δλ) * Math.cos(φ2);
      const x = Math.cos(φ1)*Math.sin(φ2) - Math.sin(φ1)*Math.cos(φ2)*Math.cos(Δλ);
      const bearing = (Math.atan2(y, x) * 180 / Math.PI + 360) % 360;

      let height = 12;
      const tags = el.tags || {};
      if (tags.height) height = parseFloat(tags.height);
      else if (tags['building:levels']) height = parseFloat(tags['building:levels']) * 3.5;
      if (isNaN(height) || height <= 0) height = 12;

      const elevAngle = Math.atan2(height, dist) * 180 / Math.PI;
      const binIndex = Math.round(bearing / 22.5) % 16;
      
      if (elevAngle > mergedAngles[binIndex].angle) {
        mergedAngles[binIndex].angle = Math.round(elevAngle * 10) / 10;
        applied++;
      }
    });
    
    console.log(`Applied ${applied} buildings`);
    return mergedAngles;
  } catch(e) {
    console.error(e);
    return horizonAngles;
  }
}

const baseAngles = Array.from({ length: 16 }, (_, i) => ({ 
  dir: ['N','NNE','NE','ENE','E','ESE','SE','SSE','S','SSW','SW','WSW','W','WNW','NW','NNW'][i], 
  bearing: i * 22.5, 
  angle: 0 
}));

// WTC area
getBuildingObstructions(40.7127, -74.0134, baseAngles).then(res => {
  console.log(res);
});
